lib/nixos-system.nix

```diff
@@ -1,43 +1,93 @@
 { self, lib, home-manager }:
 
-{ modules ? [ ], globalSpecialArgs ? { }, nixosSpecialArgs ? { }, hmSpecialArgs ? { }, ... } @ args:
-lib.nixosSystem (lib.recursiveUpdate (removeAttrs args [ "globalSpecialArgs" "nixosSpecialArgs" "hmSpecialArgs" ]) {
+with lib;
+
+{
+  # The global modules are included in both NixOS and home-manager.
+  globalModules ? [ ]
+
+  # Home-manager specific modules.
+, hmModules ? [ ]
+
+  # NixOS specific modules.
+, nixosModules ? [ ]
+
+  # The global extra arguments are included in both NixOS and home-manager.
+, globalSpecialArgs ? { }
+
+  # Home-manager specific extra arguments.
+, hmSpecialArgs ? { }
+
+  # NixOS specific extra arguments.
+, nixosSpecialArgs ? { }
+
+, ...
+} @ args:
+let
+  args' = removeAttrs args [
+    "globalModules"
+    "nixosModules"
+    "hmModules"
+
+    "globalSpecialArgs"
+    "nixosSpecialArgs"
+    "hmSpecialArgs"
+  ];
+in
+nixosSystem (recursiveUpdate args' {
   specialArgs = {
-    mode = "NixOS";
-    soxin = self;
-  } // globalSpecialArgs // nixosSpecialArgs;
+    # send home-manager down to the NixOS modules
+    inherit home-manager;
 
-  modules = modules ++ [
-    {
-      _module.args = { inherit home-manager; };
-    }
+    # the mode allows us to tell at what level we are within the modules.
+    mode = "NixOS";
 
-    self.nixosModules.soxin
+    # send soxin to all NixOS modules
+    soxin = self;
+  }
+  # include the global special arguments.
+  // globalSpecialArgs
+  # include the NixOS special arguments.
+  // nixosSpecialArgs;
 
-    home-manager.nixosModules.home-manager
-    # Required when using flakes.
-    {
+  modules =
+    # include the global modules
+    globalModules
+    # include the NixOS modules
+    ++ nixosModules
+    # include all Soxin modules
+    ++ (builtins.attrValues self.nixosModules)
+    # include all home-manager modules
+    ++ (builtins.attrValues home-manager.nixosModules)
```

Comment:
Let's maybe more specific in what we want to import. I know that we manage this project and we should check nothing is added here that we don't want, but in case home-manager adds an unwanted module, let's be sure we only include the home-manager one, instead of everything.

Reply:
done